BrainPortal/app/models/demand.rb

```diff
@@ -0,0 +1,206 @@
+   "U. S. A.", "U.K.", "U.S.", "U.S.A.", "UAE", "UK", "Ukraine",  "United Arab Emirates", "United Kingdom",    "United States", "United States of America",  "Uruguay", "US", "USA",
+   "Venezuela",  "Vietnam",  "Wales"
+  ]
+  OK_COUNTRIES_HASH = {}; OK_COUNTRIES_LIST.each { |c| OK_COUNTRIES_HASH[c.downcase]=true }
+
+
+  def strip_blanks #:nodoc:
+    [
+      :title, :first, :middle, :last,
+      :institution, :department, :position, :email,
+      :street1, :street2, :city, :province, :country, :postal_code,
+      :login, :comment
+    ].each do |att|
+      val = read_attribute(att) || ""
+      write_attribute(att, val.strip)
+    end
+    self.login = (self.login.presence || "").downcase
+    true
+  end
+
+  # Token inserted in email to new user to confirm their address
+  def generate_token
+    tok = ""
+    20.times { c=sprintf("%d",rand(10)); tok += c }
+    self.confirm_token = tok
+  end
+
+  def full #:nodoc:
+    "#{title} #{first} #{middle} #{last}".strip.gsub(/  +/, " ")
+  end
+
+  def approved? #:nodoc:
+    self.approved_by.present? && self.approved_at.present?
+  end
+
+  def is_suspicious?  # types: 1=warning, 2=weird_entries, 3=keyboard_banging
+    country = (self.country.presence || "").downcase
+    full_cat = "#{full_name}|#{institution}|#{department}|#{email}|#{country}|#{province}|#{city}|#{postal_code}|#{street1}|#{street2}|#{comment}|#{login}".downcase
+    return 3 if full_cat =~ /qwe|tyu|uio|asd|sdf|dfg|fgh|hjk|jkl|zxc|xcv|cvb|vbn|bnm/i # keyboard banging
+    return 3 if full_cat =~ /shit|fuck|cunt|blah|piss|vagina|mother|nigg|negro/i # obscenities
+    return 3 if full_cat =~ /([a-z])\1\1\1/i
+    return 2 if first.downcase == last.downcase || first.downcase == middle.downcase || middle.downcase == last.downcase
+    return 2 if first.downcase !~ /[a-z]/i || last.downcase !~ /[a-z]/i
+    return 1 unless OK_COUNTRIES_HASH[country]
+    nil
+  end
+
+  alias full_name full
+
+  def dup_email? #:nodoc:
+    User.exists?(:email => self.email)
+  end
+
+  def dup_login? #:nodoc:
+    User.exists?(:login => self.login)
+  end
+
+  # This is the method that actually creates the user in CBRAIN's database
+  def after_approval
+    #puts "Approving: #{self.full}"
+    chars = (('a'..'z').to_a + ('A'..'Z').to_a + ('0'..'9').to_a).shuffle.reject { |c| c =~ /[01OolI]/ }
+    salt  = chars[0] + chars[1]
+    password = ""
+    12.times { password += chars[rand(chars.size)] }
+
+    mylogin = self.login.presence || (self.first[0,1] + self.last).downcase
+    self.login = mylogin
+    unless self.valid?
+       raise "Current record is invalid. Probably: login name incorrect. Check form values."
+    end
+
+    if self.dup_login?
+      res = ApprovalResult.new
+      res.diagnostics = "Failed to create user, " + self.login + " already exists"
+
+      raise "Could not create user: " + self.login + " already exists in CBRAIN"
+    else
+      u = User.new
```

Comment:
Also you can put placeholders for fields that are not in the model, e.g. province:
```ruby
   #:province => self.province,
```